advent-of-code/rucksack-problem/src/rucksack_problem/core.clj

```diff
@@ -0,0 +1,136 @@
+;; Advent of Code Day 03
+
+;; Part 1
+;;      Process a single rucksack
+;;   1. Divide rucksack string into 2 halves
+;;   2. Check for the common item (only one exists) in both halves
+;;   3. Find the priority of the common item using the below relationships:
+;;      Priority of a - z: 1 - 26
+;;      Priority of A - Z: 27 - 52)
+;;
+;;      And then expand to multiple rucksacks to solve Part 1
+
+;; Lifting some input restrictions as another exercise
+;; Lift restriction of a single common item. Code generic solution where priorities are counted
+;; for each pair of common items between the 2 compartments.
+;; 1. create hashmap for compartment 1 and 2 with key: item, value: frequency of items
+;; 2. find pairs of common keys between both hashmaps
+;; 3. minimum value of each key pair is the priority multiple
+;; 4. sum of (priority * multiple) for each key pair
+
+;; Part 2
+;; For 1 group of 3 rucksacks
+;; 1. find common item in each group of 3 rucksacks.
+;; 2. find priority of common item
+;; Repeat the above for each group of 3 rucksacks, and sum the priorities of common items.
+
+(ns rucksack-problem.core
+  (:require [clojure.string :as string]
+            [clojure.set :as set])
+  (:gen-class))
+
+(defn input-file->string-seq
+  "Read file and extract strings separated by \n into a vector of strings"
+  [f]
+  (-> (slurp f)
+      (string/split #"\n")))
+
+(defn compartmentalize
+  "Divide rucksack contents into 2 compartments
+   1 string -> 2 lists of chars"
+  [rucksack-contents]
+  (let [half-length (/ (count rucksack-contents) 2)]
+    (->> rucksack-contents
+         (partition half-length))))
+
+(defn find-common-item
+  "Find the common item between 2 compartments"
+  [compartment-1, compartment-2]
+  (let [c1 (set compartment-1)
+        c2 (set compartment-2)]
+    (some #(get c2 %) c1)))
+;; can also use set intersection here
+
+(defn find-badge
+  "Find the common item (badge) between a group of 3 rucksacks.
+   Used in Part 2 of problem."
+  [rucksack-1, rucksack-2, rucksack-3]
+  (let [r1 (set rucksack-1)
+        r2 (set rucksack-2)
+        r3 (set rucksack-3)]
+    (first (set/intersection r1 r2 r3))))
```

Comment:
These two functions have some duplicated logic. They are both finding a common item in sequences of items. Consider having these delegate to a generic utility function.

Reply:
Yes, the logic is duplicated.
So `find-common-item` and `find-badge` will remain, and they would both call a new utility function within themselves. I'll give this a go.

Reply:
The new utility function will take minimum 2 + any number of collections as a rest argument. It will convert these to sets, and then return the first common element between the collections. (We only expect one common item as per the problem statement).